File: src/main/java/com/hackclub/hccore/listeners/PlayerListener.java
package com.hackclub.hccore.listeners;

import static net.kyori.adventure.text.Component.text;

import com.hackclub.hccore.HCCorePlugin;
import com.hackclub.hccore.PlayerData;
import com.hackclub.hccore.enums.Emotes;
import com.hackclub.hccore.playermessages.WelcomeMessage;
import com.hackclub.hccore.playermessages.player.BanMessage;
import com.hackclub.hccore.playermessages.player.ChatMessage;
import com.hackclub.hccore.playermessages.player.JoinMessage;
import com.hackclub.hccore.playermessages.player.LeaveMessage;
import com.hackclub.hccore.playermessages.player.MustLinkMessage;
import com.hackclub.hccore.playermessages.player.ServerFullMessage;
import io.papermc.paper.event.player.AsyncChatEvent;
import java.util.UUID;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
import org.bukkit.event.player.AsyncPlayerPreLoginEvent.Result;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerKickEvent;
import org.bukkit.event.player.PlayerKickEvent.Cause;
import org.bukkit.event.player.PlayerLoginEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerRespawnEvent;

public class PlayerListener implements Listener {

  private final HCCorePlugin plugin = HCCorePlugin.getInstance();

  @EventHandler
  public void onEntityDamage(final EntityDamageEvent event) {
    if (!(event.getEntity() instanceof Player player)) {
      return;
    }

    this.plugin.getDataManager().getData(player).setLastDamagedAt(System.currentTimeMillis());
  }

  @EventHandler
  public void onPlayerDeath(final PlayerDeathEvent event) {
    Component deathMessage = event.deathMessage();
    if (deathMessage == null) {
      return;
    }
    String message = PlainTextComponentSerializer.plainText().serialize(deathMessage);

    message = message.replace(event.getEntity().getName(),
        PlainTextComponentSerializer.plainText().serialize(event.getEntity().displayName()));

    Player killer = event.getEntity().getKiller();
    if (killer != null) {
      message = message.replace(killer.getName(),
          PlainTextComponentSerializer.plainText().serialize(killer.displayName()));
    }

    event.deathMessage(text(message));
  }

  @EventHandler(priority = EventPriority.LOWEST)
  public void onAsyncChat(final AsyncChatEvent event) {
    event.renderer((source, sourceDisplayName, message, viewer) -> {
      PlayerData data = this.plugin.getDataManager().getData(source);
      TextColor messageColor = data.getMessageColor();
//    TextColor nameColor = data.getNameColor();

      String rawMessage = ((TextComponent) event.message()).content();

      // Apply the player's chat color to the message and translate color codes

      Component nameComponent = source.displayName().hoverEvent(
          net.kyori.adventure.text.event.HoverEvent.showEntity(source.getType(), source.getUniqueId(),
              source.name()));

      return ChatMessage.get(nameComponent, Emotes.parseString(rawMessage), messageColor);
    });
  }

  public static boolean isSlackJoinAllowed(PlayerData data) {
    HCCorePlugin plugin = HCCorePlugin.getPlugin(HCCorePlugin.class);

    if (!plugin.getConfig()
        .getBoolean("settings.slack-link.enabled", false)) {
      return true;
    }
    if (!plugin.getConfig()
        .getBoolean("settings.slack-link.required", false)) {
      return true;
    }

    String id = data.getSlackId();

    return !plugin.getSlackBot().isDeactivated(id);
  }


  @EventHandler(priority = EventPriority.LOWEST) // Runs foremost
  public void onPrePlayerLogin(final AsyncPlayerPreLoginEvent event) {
    UUID playerUUID = event.getUniqueId();
    PlayerData data = this.plugin.getDataManager()
        .getData(Bukkit.getServer().getOfflinePlayer(playerUUID));
    data.load();

    if (!isSlackJoinAllowed(data)) {
      // Check for slack link
      this.plugin.getLogger()
          .info("Preventing " + event.getName() + "'s join because they are not linked");
      Component kickMessage = getSlackLinkMessage(playerUUID);
      event.disallow(Result.KICK_WHITELIST, kickMessage);
    }
  }

  public static Component getSlackLinkMessage(UUID playerUUID) {
    String code = HCCorePlugin.getPlugin(HCCorePlugin.class).getSlackBot()
        .generateVerificationCode(playerUUID);
    int codeExpires = HCCorePlugin.getPlugin(HCCorePlugin.class).getConfig()
        .getInt("settings.slack-link.link-code-expiration", 60 * 10);
    String baseCommand =
        (String)HCCorePlugin.getInstance().getConfig().get("settings.slack-link.base-command", "minecraft");
    return MustLinkMessage.get(baseCommand, code, codeExpires);
  }

  @EventHandler(priority = EventPriority.MONITOR)
  public void onJoin(final PlayerJoinEvent event) {
    Player player = event.getPlayer();
    this.plugin.getDataManager().registerPlayer(player);
    this.plugin.getDataManager().getData(player)
        .setLastActiveAt(System.currentTimeMillis());

    // NOTE: Title isn't cleared when the player leaves the server
    player.clearTitle();
    event.joinMessage(
        JoinMessage.get(player.displayName().hoverEvent(event.getPlayer())));
    plugin.advancementTab.showTab(player);
  }

  @EventHandler
  public void onFirstJoin(final PlayerJoinEvent joinEvent) {
    Player player = joinEvent.getPlayer();
    if (player.hasPlayedBefore()) {
      return;
    }
    player.sendMessage(WelcomeMessage.get());
  }

  @EventHandler(priority = EventPriority.HIGHEST)
  public void onFailedLogin(final PlayerLoginEvent event) {
    if (event.getResult() == PlayerLoginEvent.Result.ALLOWED) {
      return;
    }

    Component message;
    switch (event.getResult()) {
      case KICK_BANNED -> message = BanMessage.get(event.getPlayer().getUniqueId().toString());
      case KICK_FULL -> message = ServerFullMessage.get();
      default -> message = event.kickMessage();
    }
    event.kickMessage(message);
  }

  @EventHandler
  public void onPlayerKick(final PlayerKickEvent event) {
    Component message;
    if (event.getCause() == Cause.BANNED) {
      message = BanMessage.get(event.getPlayer().getUniqueId().toString());
    } else {
      message = event.reason();
    }

    event.reason(message);
  }

  @EventHandler
  public void onPlayerRespawn(final PlayerRespawnEvent event) {
    // Reset lastDamagedAt
    this.plugin.getDataManager().getData(event.getPlayer()).setLastDamagedAt(0);
  }

  @EventHandler(priority = EventPriority.MONITOR) // Runs very last
  public void onPlayerQuit(final PlayerQuitEvent event) {
    // NOTE: Title isn't cleared when the player leaves the server
    // event.getPlayer().resetTitle();
    event.quitMessage(
        LeaveMessage.get(
            event.getPlayer().displayName().hoverEvent(event.getPlayer())));

    this.plugin.getDataManager().unregisterPlayer(event.getPlayer());
  }
}

File: src/main/java/com/hackclub/hccore/listeners/PlayerListener.kt
package com.hackclub.hccore.listenersimport

import net.md_5.bungee.api.chat.ComponentBuilder
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.*

com.hackclub.hccore.HCCorePlugin
import com.hackclub.hccore.PlayerData
import com.hackclub.hccore.tasks.CheckAdAstraTask
import com.hackclub.hccore.utils.gson.GsonUtil
import com.hackclub.hccore.utils.gson.LocationSerializer
import com.hackclub.hccore.utils.gson.LocationMapDeserializer
import com.hackclub.hccore.utils.gson.LocationMapSerializer
import com.hackclub.hccore.utils.gson.LocationDeserializer
import java.util.LinkedHashMap
import com.hackclub.hccore.events.player.PlayerAFKStatusChangeEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.events.ReadyEvent
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.JDABuilder
import javax.security.auth.login.LoginException
import java.lang.Exception
import com.hackclub.hccore.commands.ColorCommand
import java.util.stream.Collectors
import com.hackclub.hccore.commands.StatsCommand
import java.text.SimpleDateFormat
import com.hackclub.hccore.commands.UpvoteCommand
import com.hackclub.hccore.commands.DownvoteCommand
import com.hackclub.hccore.commands.TableflipCommand
import java.lang.Runnable
import com.hackclub.hccore.listeners.SleepListener
import com.comphenix.protocol.events.ListenerPriority
import com.comphenix.protocol.PacketType
import com.comphenix.protocol.events.PacketAdapter
import com.comphenix.protocol.events.PacketEvent
import com.comphenix.protocol.wrappers.EnumWrappers
import com.comphenix.protocol.wrappers.WrappedGameProfile
import org.bukkit.craftbukkit.v1_17_R1.advancement.CraftAdvancement
import net.minecraft.advancements.AdvancementDisplay
import net.minecraft.network.chat.IChatBaseComponent
import net.minecraft.advancements.AdvancementFrameType
import de.tr7zw.changeme.nbtapi.NBTTileEntity
import java.io.FileReader
import java.io.IOException
import java.io.FileWriter
import java.util.UUID
import java.util.HashMap
import com.hackclub.hccore.DataManager
import com.comphenix.protocol.ProtocolManager
import com.hackclub.hccore.discord.DiscordBot
import com.comphenix.protocol.ProtocolLibrary
import com.hackclub.hccore.commands.AFKCommand
import com.hackclub.hccore.commands.LocCommand
import com.hackclub.hccore.commands.NickCommand
import com.hackclub.hccore.commands.PingCommand
import com.hackclub.hccore.commands.ShrugCommand
import com.hackclub.hccore.commands.SpawnCommand
import com.hackclub.hccore.commands.DiscordCommand
import com.hackclub.hccore.listeners.AdvancementListener
import com.hackclub.hccore.listeners.AFKListener
import com.hackclub.hccore.listeners.BeehiveInteractionListener
import com.hackclub.hccore.listeners.PlayerListener
import com.hackclub.hccore.listeners.NameChangeListener
import com.hackclub.hccore.tasks.AutoAFKTask
import hu.trigary.advancementcreator.AdvancementFactory
import hu.trigary.advancementcreator.shared.ItemObject

class PlayerListener(plugin: HCCorePlugin) : Listener {
    private val plugin: HCCorePlugin
    @EventHandler
    fun onEntityDamage(event: EntityDamageEvent) {
        if (event.entity !is Player) {
            return
        }
        val player = event.entity as Player
        plugin.getDataManager().getData(player).setLastDamagedAt(System.currentTimeMillis())
    }

    @EventHandler
    fun onPlayerDeath(event: PlayerDeathEvent) {
        var message = event.deathMessage
        message = message!!.replace(event.entity.name,
                ChatColor.stripColor(event.entity.displayName)!!)
        val killer = event.entity.killer
        if (killer != null) {
            message = message.replace(killer.name,
                    ChatColor.stripColor(killer.displayName)!!)
        }
        event.deathMessage = message
    }

    @EventHandler
    fun onAsyncPlayerChat(event: AsyncPlayerChatEvent) {

        // Apply the player's chat color to the message and translate color codes
        val data: PlayerData = plugin.getDataManager().getData(event.player)
        val messageColor = data.messageColor.asBungee()
        val nameColor = data.nameColor.asBungee()
        val nameComponent = TextComponent(event.player.displayName)
        nameComponent.color = nameColor
        nameComponent.hoverEvent = HoverEvent(HoverEvent.Action.SHOW_TEXT,
                ComponentBuilder(event.player.name).create())
        val arrowComponent = TextComponent(" » ")
        arrowComponent.color = ChatColor.GOLD.asBungee()
        val playerChatComponent = TextComponent(ChatColor.translateAlternateColorCodes('&', event.message))
        playerChatComponent.color = messageColor
        plugin.getServer().spigot().broadcast(nameComponent, arrowComponent,
                playerChatComponent)
        event.isCancelled = true
    }

    @EventHandler(priority = EventPriority.LOWEST) // Runs foremost
    fun onPlayerJoin(event: PlayerJoinEvent) {
        plugin.getDataManager().registerPlayer(event.player)
        // Set the initial active time
        plugin.getDataManager().getData(event.player)
                .setLastActiveAt(System.currentTimeMillis())

        // NOTE: Title isn't cleared when the player leaves the server
        event.player.resetTitle()
        event.joinMessage = ChatColor.YELLOW
                .toString() + ChatColor.stripColor(event.player.displayName) + " joined the game"
    }

    @EventHandler
    fun onPlayerLogin(event: PlayerLoginEvent) {
        if (event.result == PlayerLoginEvent.Result.ALLOWED
                || event.result == PlayerLoginEvent.Result.KICK_OTHER) {
            return
        }
        var message: String? = null
        when (event.result) {
            PlayerLoginEvent.Result.KICK_BANNED -> message = """
     ${ChatColor.RED}${ChatColor.BOLD}You’ve been banned :(
     
     ${ChatColor.RESET}${ChatColor.WHITE}If you believe this was a mistake, please DM ${ChatColor.AQUA}@alx or @eli ${ChatColor.WHITE}on Slack.
     """.trimIndent()
            PlayerLoginEvent.Result.KICK_FULL -> message = """
     ${ChatColor.RED}${ChatColor.BOLD}The server is full!
     
     ${ChatColor.RESET}${ChatColor.WHITE}Sorry, it looks like there’s no more room. Please try again in ~20 minutes.
     """.trimIndent()
            PlayerLoginEvent.Result.KICK_WHITELIST -> message = """
     ${ChatColor.RED}${ChatColor.BOLD}You’re not whitelisted!
     
     ${ChatColor.RESET}${ChatColor.WHITE}Join ${ChatColor.AQUA}#minecraft ${ChatColor.WHITE}on Slack and ping ${ChatColor.AQUA}@alx or @eli ${ChatColor.WHITE}to be added.
     """.trimIndent()
            else -> {
            }
        }
        event.kickMessage = message!!
    }

    @EventHandler
    fun onPlayerRespawn(event: PlayerRespawnEvent) {
        // Reset lastDamagedAt
        plugin.getDataManager().getData(event.player).setLastDamagedAt(0)
    }

    @EventHandler(priority = EventPriority.MONITOR) // Runs very last
    fun onPlayerQuit(event: PlayerQuitEvent) {
        // NOTE: Title isn't cleared when the player leaves the server
        // event.getPlayer().resetTitle();
        event.quitMessage = ChatColor.YELLOW
                .toString() + ChatColor.stripColor(event.player.displayName) + " left the game"
        plugin.getDataManager().unregisterPlayer(event.player)
    }

    init {
        this.plugin = plugin
    }
}